### ISSUE TITLE
ExamReview #5

### ISSUE THREAD

**Issue by rayshroom** [seneca-btp500f19/btp500-f19#15]
5. If you were to implement a reverse phone directory (You enter a phone number to find a person), what data structure would you use for the table.

A: Depends on the purpose of the use case. If we want to know about existence of a number in the directory, a hash table would be good. If we want to know all numbers in the 647 or 905 area, a trie would be better.
-----------------
Actually shouldn't trie be better than hashtable in both retrieval and space efficiency?

**Comment by drwm-base** [comment]
Well the question asks specifically that we want to search for a person with the given phone number. 

A TRie is typically used on strings, so using it for a normal phone directory lookup would work really well (autocomplete your searches for a person). However, a TRie would need to use a hash function to allow you to search by phone number, and this could cause collisions (which would have to be dealt with anyway). A TRie is also less space efficient than a hash table, as while a hash table only uses what it needs, a TRie instantiates 26 nullptrs on every node. 

So I would say in that case your best option is a hash table, probably using linear probing and a max load factor to reduce memory and time costs.

**Comment by Grommers00** [comment]
I would say a TRie would be more effective, because a phone directory would have 100,000+ people I would assume.  
Given the size number, a hash table would constantly have to be re-sized, or re-used, and bucketting would come into play. Slowing down things even more, especially because so many numbers have the same Area Code. 

Where as a TRi would only have 10 nodes 0 to 9, and each node that would complete someones number would have to store the name of the individual whose number it would be storing. 

This would mean (assuming that a phone number is 9 digits long, and doesn't include international), the most you would need to look through is 9 nodes to get the name of your individual, and you because you would be passing in a number it would be quick and easy as you know exactly what location each node is at because of the phone number. 

Hash table should theoretically take much longer..unless your phone directory was incredibly small.  However for the sake of this question, I'm going to assume it means like a phone book.

**Comment by rayshroom** [comment]
> A TRie is typically used on strings

Phone numbers are stored as strings


> However, a TRie would need to use a hash function to generate a string key

You don't need a key for TRie nodes

**Comment by drwm-base** [comment]
Right but to my understanding a TRie is alphabetical only. If it is not alphabetical (ie 26 nodes for the alphabet), then I don't know if it is a TRie anymore. Yeah, the best solution by far is to have a TRie - like structure with nodes 0-9 that represent each number in a phone number combination. I'm just not sure if that is a TRie anymore, hence why I went with hash table. Also @Grommers00 you would still have to account for bucketing as two people can share a phone number if they live at the same house.

It's definitely a tree, but is it a TRie?

@rayshroom that is why I said it needs a hash function, to generate an alpha string key to use the TRie.

**Comment by Grommers00** [comment]
@drwm-base yeah, good point - I'm more of the assumption that one name is linked to one phone number, but definitely would be the case e.g. work numbers (as extensions would have to be stored seperately)

**Comment by cathyatseneca** [comment]
you can translate the idea of a phone number into a trie... I mean text is arbitrary and its easy enough to find any digit of a number with / and %. extensions are just more digits...

**Comment by drwm-base** [comment]
Can a TRie be non-alphabetical?

**Comment by saketepe** [comment]
Not sure, but i assume that all it needs is to have is a node that points to other node that hold a set of all possible  combinations.
exp
root
----------                                                        l
l var[num]                                                       l
l arr[*] // points to a list of 10 node pointers    l 
---------                                                         l

I think as long as the nodes look something like this its a try.
